src/typechecker.ts

```diff
@@ -298,3 +357,397 @@ function textFromPreviousPart(seq: Seq, index: number): string | null {
+  | { kind: 'list'; of: Type }
+  | { kind: 'record' } // TODO more precision
+  | { kind: 'completion'; of: Type }
+  | { kind: 'real' }
+  | { kind: 'integer' }
+  | { kind: 'non-negative integer' }
+  | { kind: 'concrete real'; value: string }
+  | { kind: 'ES value' }
+  | { kind: 'string' }
+  | { kind: 'number' }
+  | { kind: 'integral number' }
+  | { kind: 'bigint' }
+  | { kind: 'null' }
+  | { kind: 'concrete string'; value: string }
+  | { kind: 'concrete number'; value: string }
+  | { kind: 'concrete bigint'; value: string }
+  | { kind: 'enum value'; value: string };
+
+type NonUnion = Exclude<Type, { kind: 'union' }>;
+
+const simpleKinds = new Set<Type['kind']>([
+  'unknown',
+  'never',
+  'record',
+  'real',
+  'integer',
+  'non-negative integer',
+  'ES value',
+  'string',
+  'number',
+  'integral number',
+  'bigint',
+]);
+
+const dominateGraph: Partial<Record<Type['kind'], Type['kind'][]>> = {
+  // @ts-expect-error TS does not know about __proto__
+  __proto__: null,
+  record: ['completion'],
+  real: ['integer', 'non-negative integer', 'concrete real'],
+  integer: ['non-negative integer'],
+  'ES value': [
+    'string',
+    'number',
+    'integral number',
+    'bigint',
+    'concrete string',
+    'concrete number',
+    'concrete bigint',
+  ],
+  string: ['concrete string'],
+  number: ['integral number', 'concrete number'],
+  bigint: ['concrete bigint'],
+};
+
+/*
+The type lattice used here is very simple (aside from explicit unions).
+As such we really only need to define the `dominates` relationship and apply trivial rules:
+if `x` dominates `y`, then `join(x,y) = x` and `meet(x,y) = y`; if neither dominates the other than the join is top and the meet is bottom.
+Unions take a little more work.
+*/
+
+function dominates(a: Type, b: Type): boolean {
+  if (a.kind === 'unknown' || b.kind === 'never') {
+    return true;
+  }
+  if (b.kind === 'union') {
+    return b.of.every(t => dominates(a, t));
+  }
+  if (a.kind === 'union') {
+    // not necessarily true for arbitrary lattices, but true for ours
+    return a.of.some(t => dominates(t, b));
+  }
+  if (
+    (a.kind === 'list' && b.kind === 'list') ||
+    (a.kind === 'completion' && b.kind === 'completion')
+  ) {
+    return dominates(a.of, b.of);
+  }
+  if (simpleKinds.has(a.kind) && simpleKinds.has(b.kind) && a.kind === b.kind) {
+    return true;
+  }
+  if (dominateGraph[a.kind]?.includes(b.kind) ?? false) {
+    return true;
+  }
+  if (
+    (a.kind === 'integer' && b.kind === 'concrete real') ||
+    (a.kind === 'integral number' && b.kind === 'concrete number')
+  ) {
+    return !b.value.includes('.');
+  }
+  if (a.kind === 'non-negative integer' && b.kind === 'concrete real') {
+    return !b.value.includes('.') && b.value[0] !== '-';
+  }
+  if (
+    a.kind === b.kind &&
+    ['concrete string', 'concrete number', 'concrete bigint', 'enum value'].includes(a.kind)
+  ) {
+    // @ts-expect-error TS is not quite smart enough for this
```

Comment:
There's a couple ways that you could avoid this. Either use some predicates that narrow `a` or a big disjunction.

Reply:
Yeah but I don't want to write more complicated code just to please the type checker, since this code is in fact clearly correct. That's what `@ts-expect-error` is for.